ocaml/xapi/xapi_vm.ml

```diff
@@ -1611,3 +1611,13 @@ let set_NVRAM_EFI_variables ~__context ~self ~value =
       let value = (key, value) :: List.remove_assoc key nvram in
       Db.VM.set_NVRAM ~__context ~self ~value
   )
+
+let restart_device_models ~__context ~self =
+  let host = Db.VM.get_resident_on ~__context ~self in
+  (* As it is implemented as a localhost migration, just reuse message
+   * forwarding of "pool_migrate" to handle "allowed operation" and "message
+   * routing" *)
+  Helpers.call_api_functions ~__context (fun rpc session_id ->
+      Client.VM.pool_migrate ~rpc ~session_id ~vm:self ~host
```

Comment:
Which error will be returned if the VM is not running? It should be a bad power state.

Reply:
The handler of "pool_migrate" had handled kinds of errors including the "bad power state", here is the tested result:

```
[root@lcy2-dt44 ~]# xe vm-restart-device-models vm=rhel8_202
You attempted an operation on a VM that was not in an appropriate power state at the time; for example, you attempted to start a VM that was already running. The parameters returned are the VM's handle, and the expected and actual VM state at the time of the call.
vm: 698f91f3-84f8-34e3-714e-f4cd7d8a1207 (rhel8_202)
expected: running
actual: halted
```

Reply:
That's good. I was worried that it would first complain about the `host` param being `OpaqueRef:NULL` (from `resident_on`).